R/deps_installation_proposal.R

```diff
@@ -299,6 +302,56 @@ get_refs_from_desc <- function(d) {
   res[res_idx]
 }
 
+#' Replace Remotes in the `desc` that have been resolved to a GitHub tag or are
+#' in CRAN
+#'
+#' Replaces any existing Remotes entry with the resolved GitHub tag from the
+#' `new_refs`.
+#'
+#' It keeps all the existing Remotes that have not been resolved in `new_refs`.
+#'
+#' @param d (`desc`) DESCRIPTION object
+#' @param new_refs (`list`) remote references that have been resolved and are
+#' being updated in `Config/Needs/verdepcheck`
+#' @keywords internal
+desc_remotes_cleanup <- function(d, new_refs) {
+  # Parse the remotes to retrieve the package names
+  remotes <- pkgdepends::parse_pkg_refs(d$get_remotes())
+
+  # Get the packages defined in remotes
+  #  (making sure that only packages that are already defined here are modified)
+  remotes_pkg <- vapply(remotes, `[[`, character(1), "package")
+
+  # Find which packages of the new_refs are defined in Remotes
+  new_refs_gh <- vapply(
+    new_refs,
+    function(.x) {
+      isTRUE(.x$package %in% remotes_pkg) && inherits(.x, "remote_ref_github")
+    },
+    logical(1)
+  )
```

Comment:
here it would be more efficient to make use of `Filter()` as opposed to build a mask and then apply it on the vector.

Reply:
`vapply` seems a bit faster, but they're pretty neck in neck.

I do like the clean code of using `Filter`!

```R
Browse[1]> .f <- function(.x) isTRUE(.x$package %in% remotes_pkg) && inherits(.x, "remote_ref_github")

Browse[1]> microbenchmark::microbenchmark(
    times = 100L,
    check = "equal",
    filter = Filter(.f, new_refs),
    vapply = new_refs[vapply(new_refs, .f, logical(1))],
    setup = new_refs[] # to remove variability on first call
  )
#Unit: microseconds
#   expr    min      lq     mean  median      uq    max neval
# filter 49.123 49.7765 51.15354 50.1840 51.0315 83.586   100
# vapply 45.296 46.1205 46.99391 46.6455 47.2345 57.032   100

Browse[1]> microbenchmark::microbenchmark(times = 100L, vapply = new_refs[vapply(new_refs, .f, logical(1))])
#Unit: microseconds
#   expr    min     lq     mean  median      uq    max neval
# vapply 44.908 46.119 49.94768 47.6345 53.4715 89.241   100
  
Browse[1]> microbenchmark::microbenchmark(times = 100L, filter = Filter(.f, new_refs))
#Unit: microseconds
#   expr    min      lq     mean median     uq    max neval
# filter 49.841 50.5585 53.84665 51.382 56.061 133.48   100
```

Reply:
:heavy_check_mark:  Done in ac12fea